### PR TITLE
2110 generic channels

### PR DESCRIPTION
**Pull request by Baron-Severin**
### Pull Request checklist

- [ ] This PR includes thorough **tests** or an explanation of why it does not
- [x] This PR includes a **CHANGELOG entry** or does not need one
- [ ] I have considered adding **QA labels** on the associated issue (not this PR; `qa-ready` or `qa-denied`)

### CODE REVIEW

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+
+        // DiffUtil allows diff calculation between two lists and output a list of update
+        // operations that converts the first list into the second one
+        val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
+            override fun getOldListSize(): Int = tiles.size
+
+            override fun getNewListSize(): Int = newTiles.size
+
+            override fun areItemsTheSame(oldPos: Int, newPos: Int): Boolean {
+                return tiles[oldPos].url == newTiles[newPos].url
```
Comment (mcomella) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt]: Are urls really considered unique ids? I feel like we should just do the same thing as `areContentsTheSame` to be safe.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    fun setTiles(newTiles: List<ChannelTile>) {
+        if (itemCount == 0) {
+            this.tiles = newTiles
+            notifyDataSetChanged()
+            return
+        }
+
+        // DiffUtil allows diff calculation between two lists and output a list of update
+        // operations that converts the first list into the second one
+        val diff = DiffUtil.calculateDiff(object : DiffUtil.Callback() {
```
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt]: I'm not familiar with it, but I'll take a look.

File: app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTile.kt [hunk 3]
```diff
@@ -75,11 +76,16 @@ class CustomPinnedTile(
...
-    override fun toChannelTile(screenshotStoreWrapper: ScreenshotStoreWrapper): ChannelTile {
+    override fun toChannelTile(imageUtilityWrapper: PinnedTileImageUtilWrapper): ChannelTile {
+        val backup = imageUtilityWrapper.generatePinnedTilePlaceholder(url)
+
         return ChannelTile(
                 url = url,
                 title = title,
-                setImage = { view -> PicassoWrapper.client.load(screenshotStoreWrapper.getFileForUUID(id)).into(view) } // todo: fix scope, double check this is okay.
+                setImage = { view -> PicassoWrapper.client
+                        .load(imageUtilityWrapper.getFileForUUID(id))
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTile.kt]: Does this throw anything that doesn't get caught if the UUID doesn't exist as a file? (Maybe this is already handlede before)

File: app/src/main/java/org/mozilla/tv/firefox/utils/FormattedDomainWrapper.kt [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+
+package org.mozilla.tv.firefox.utils
+
+import android.app.Application
+import java.net.URI
+
+/**
+ * TODO
+ */
+class FormattedDomainWrapper(private val application: Application) {
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/utils/FormattedDomainWrapper.kt]: Unrelated to your changes, what do you think about this being called ReadableDomain? Formatted sounds so unhelpful.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 5]
```diff
@@ -291,6 +291,36 @@ class NavigationOverlayFragment : Fragment() {
...
     }
 
+    private fun createChannelFactory(): DefaultChannelFactory = DefaultChannelFactory(
+            loadUrl = { urlStr ->
+                if (urlStr.isNotEmpty()) {
+                    onNavigationEvent.invoke(NavigationEvent.LOAD_TILE, urlStr, null)
+                }
+            },
+            onTileLongClick = openHomeTileContextMenu,
+            onTileFocused = {
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Not sure if this should be in the Factory, if PinnedTiles are the only ones that can be unpinned.

Comment (liuche) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Same with long click.

Comment (liuche) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Though maybe that's fine, if it's already here. Who knows what might change with our future tiles (like dismissing Pocket tiles, the way Desktop does it).

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayViewModel.kt [hunk 6]
```diff
@@ -26,4 +36,17 @@ class NavigationOverlayViewModel(
     val viewIsSplit: LiveData<Boolean> = sessionRepo.legacyState.map {
         it.currentUrl != URLs.APP_URL_HOME
     }
+
+    val pinnedTiles: Observable<ChannelDetails> = pinnedTileRepo.pinnedTiles
+            .observeOn(Schedulers.computation())
+            .map { it.values.map { it.toChannelTile(imageUtilityWrapper, formattedDomainWrapper) } }
+            .map { ChannelDetails(title = "Pinned Tiles", tileList = it) } // TODO extract string
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayViewModel.kt]: We'll need a new string for this actually, we should run this by Meridel (or her stand-in).

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+            }
+        })
+
+        tiles = newTiles
+        diff.dispatchUpdatesTo(this)
+    }
+
+    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): DefaultChannelTileViewHolder {
+        val inflater = LayoutInflater.from(parent.context)
+//        val view = inflater.inflate(R.layout.default_channel_tile, parent, false)
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt]: ^

File: app/src/main/res/layout/fragment_navigation_overlay.xml [hunk 8]
```diff
@@ -99,14 +99,14 @@
...
 
-            <androidx.recyclerview.widget.RecyclerView
-                android:id="@+id/tileContainer"
-                android:layout_width="match_parent"
-                android:layout_height="wrap_content"
-                android:layout_marginStart="45dp"
-                android:layout_marginEnd="45dp"
-                android:clipChildren="false"
-                android:clipToPadding="false" />
+            <!--<androidx.recyclerview.widget.RecyclerView-->
```
Comment (liuche) [inline app/src/main/res/layout/fragment_navigation_overlay.xml]: Just delete it?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelFactory.kt [hunk 9]
```diff
@@ -0,0 +1,42 @@
+// todo: license
```
Comment (liuche) [inline app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelFactory.kt]: ^

File: app/src/test/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileTest.kt [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+                formattedDomainWrapper,
+                pinnedTileRepo
+        )
+        testObserver = overlayVm.pinnedTiles.test()
+    }
+
+    @Test
+    fun `WHEN repo emits a successful load THEN view model should emit a list of same number of tiles`() {
+        assertEquals(1, testObserver.valueCount())
+        assertEquals(10, testObserver.values().last().tileList.size)
```
Comment (liuche) [inline app/src/test/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileTest.kt]: Oh for a second I was confused, but this 10 is the number of default pinned tiles that we ship already. Can you add a comment here?

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/channels/DefaultChannelAdapter.kt]: Neat, that cut out a good chunk of code.  Thanks!

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 11]
```diff
@@ -135,11 +133,11 @@ class NavigationOverlayFragment : Fragment() {
     private lateinit var serviceLocator: ServiceLocator
     private lateinit var navigationOverlayViewModel: NavigationOverlayViewModel
     private lateinit var toolbarViewModel: ToolbarViewModel
-    private lateinit var pinnedTileViewModel: PinnedTileViewModel
     private lateinit var pocketViewModel: PocketViewModel
     private lateinit var hintViewModel: HintViewModel
+    private lateinit var pinnedTileChannel: DefaultChannel
 
-    private var tileAdapter: PinnedTileAdapter? = null
+//    private var tileAdapter: PinnedTileAdapter? = null
```
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Whoops

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt [hunk 12]
```diff
@@ -135,11 +133,11 @@ class NavigationOverlayFragment : Fragment() {
     private lateinit var serviceLocator: ServiceLocator
     private lateinit var navigationOverlayViewModel: NavigationOverlayViewModel
     private lateinit var toolbarViewModel: ToolbarViewModel
-    private lateinit var pinnedTileViewModel: PinnedTileViewModel
     private lateinit var pocketViewModel: PocketViewModel
     private lateinit var hintViewModel: HintViewModel
+    private lateinit var pinnedTileChannel: DefaultChannel
```
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: Hmm, I've seen them it a singular channel, made up of plural tiles.  I could change the observe methods to `observePinnedTileChannel`, if you would prefer?

File: app/src/main/res/layout/fragment_navigation_overlay_orig.xml [hunk 13]
```diff
@@ -92,6 +92,14 @@
                     android:focusable="true"
                     android:nextFocusUp="@id/navUrlInput" />
 
+                <LinearLayout
+                    android:id="@+id/channelContainer"
```
[the remark this answers is not in the document]
Comment (Baron-Severin) [reply app/src/main/res/layout/fragment_navigation_overlay_orig.xml]: Multiple, I'll rename it

File: app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTile.kt (hunk 3 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTile.kt]: Just tested, it cleanly falls back to the default.  Good callout.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt (hunk 5 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayFragment.kt]: I'm unsure on this, tbh.  I timeboxed trying to find a good removal solution, and didn't manage to.  I have that as part of the follow up.

File: app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayViewModel.kt (hunk 6 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/navigationoverlay/NavigationOverlayViewModel.kt]: I put a TODO about that and _still_ forgot.  Good catch!

File: app/src/test/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileTest.kt (hunk 10 above)
Comment (Baron-Severin) [reply app/src/test/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileTest.kt]: Replaced with a constant.

File: app/src/main/java/org/mozilla/tv/firefox/utils/FormattedDomainWrapper.kt (hunk 4 above)
Comment (Baron-Severin) [reply app/src/main/java/org/mozilla/tv/firefox/utils/FormattedDomainWrapper.kt]: Hm.  I looked at the implementation, and although (as far as I know) we only use this to make it readable, the methods themselves are a lot more flexible than that.  I think I'd prefer to leave it as is, even though it's not a super descriptive name.

**Pull request merged**